### DISCUSSION TITLE
Investigate testing for the Kakarot EVM

### DISCUSSION THREAD

**Discussion by bajpai244** [kkrt-labs/kakarot#606] [General]
Let's investigate the best testing for Kakarot ZKEVM implementation, and what types of tests can we do to guarantee the integrity of the EVM.

**Comment by jobez** [comment]
Just to do a tour of what we already have:

- Unit tests:  https://github.com/kkrt-labs/kakarot/tree/main/tests/src/kakarot
- Integration tests: https://github.com/kkrt-labs/kakarot/tree/main/tests/integration/solidity_contracts
  - per opcode: https://github.com/kkrt-labs/kakarot/blob/main/tests/integration/bytecodes.py
     -  question in telegram about origination of the opcode cases: https://t.me/KakarotZkEvm/9068
        -  clement's answer: sorry, we just wrote most of them, some were also copy/pasted from evm.codes


i looked into examples of projects using https://github.com/ethereum/tests, in particular the kevm

- https://discord.com/channels/824582698147905556/824582698147905559/1087844729460379668 question in their discord
- answer:

> In this pull request, for example, we update to the paris/merge schedule, and it enables more conformance tests for the new opcode PREVRANDAO, which then must be passed by the semantics to maintain conformance:  https://github.com/runtimeverification/evm-semantics/pull/1647 

- the specific entrypoint in kevm that uses the described test cases in ethereum/test: 
   - https://github.com/runtimeverification/evm-semantics/blob/284ae464fb96036d82809eae32a80a785d873d5f/Makefile#L583
